src/app/year-calendar/year-calendar.component.ts

```diff
@@ -0,0 +1,254 @@
+  }
+
+  constructor(public yearCalendarService: YearCalendarService,
+              public utilsService: UtilsService) {
+  }
+
+  ngOnInit() {
+    this.isInited = true;
+    this.init();
+    this.initValidators();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (this.isInited) {
+      const {minDate, maxDate} = changes;
+      this.init();
+
+      if (minDate || maxDate) {
+        this.initValidators();
+      }
+    }
+  }
+
+  init() {
+    this.componentConfig = this.yearCalendarService.getConfig(this.config);
+    this.selected = this.selected || [];
+    this.currentDateView = this.displayDate
+      ? this.displayDate
+      : this.utilsService
+        .getDefaultDisplayDate(
+          this.currentDateView,
+          this.selected,
+          this.componentConfig.allowMultiSelect,
+          this.componentConfig.min
+        );
+    this.inputValueType = this.utilsService.getInputType(this.inputValue, this.componentConfig.allowMultiSelect);
+    this._shouldShowCurrent = this.shouldShowCurrent();
+  }
+
+  writeValue(value: CalendarValue): void {
+    this.inputValue = value;
+
+    if (value) {
+      this.selected = this.utilsService
+        .convertToMomentArray(value, this.componentConfig.format, this.componentConfig.allowMultiSelect);
+      this.yearMatrix = this.yearCalendarService
+        .generateYear(this.componentConfig, this.currentDateView, this.selected);
+      this.inputValueType = this.utilsService.getInputType(this.inputValue, this.componentConfig.allowMultiSelect);
+    }
+  }
+
+  registerOnChange(fn: any): void {
+    this.onChangeCallback = fn;
+  }
+
+  onChangeCallback(_: any) {
+  };
+
+  registerOnTouched(fn: any): void {
+  }
+
+  validate(formControl: FormControl): ValidationErrors | any {
+    if (this.minDate || this.maxDate) {
+      return this.validateFn(formControl.value);
+    } else {
+      return () => null;
+    }
+  }
+
+  processOnChangeCallback(value: Moment[]): CalendarValue {
+    return this.utilsService.convertFromMomentArray(
+      this.componentConfig.format,
+      value,
+      this.componentConfig.returnedValueType || this.inputValueType
+    );
+  }
+
+  initValidators() {
+    this.validateFn = this.validateFn = this.utilsService.createValidator(
+      {minDate: this.minDate, maxDate: this.maxDate},
+      this.componentConfig.format,
+      'year');
+
+    this.onChangeCallback(this.processOnChangeCallback(this.selected));
+  }
+
+  yearClicked(year: IYear) {
+    this.selected = this.utilsService
+      .updateSelected(this.componentConfig.allowMultiSelect, this.selected, year, 'year');
+    this.yearMatrix = this.yearCalendarService
+      .generateYear(this.componentConfig, this.currentDateView, this.selected);
+    this.onSelect.emit(year);
+  }
+
+  onLeftNav() {
+    this.currentDateView = this.currentDateView.clone().subtract(12, 'year');
+    this.yearMatrix = this.yearCalendarService.generateYear(this.componentConfig, this.currentDateView, this.selected);
```

Comment:
should be changed when currentDateView is changed, please see other components

Reply:
Isn't this equivalent to the same method, in `month-calendar.component.ts`? I don't understand your comment.

Reply:
take a look at https://github.com/vlio20/angular-datepicker/blob/master/src/app/month-calendar/month-calendar.component.ts#L86

Reply:
Yes, and my `currentDateView` method in `year-calendar.component.ts` does the same (please note I renamed the `this.yearMonths` to `this.yearMatrix`, for consistency): 
<snip>
    set currentDateView(current: Moment) {
        this._currentDateView = current.clone();
        this.yearMatrix = this.yearCalendarService
          .generateYear(this.componentConfig, this._currentDateView, this.selected);
</snip>

Reply:
(didn't find a multi-line code formatting)